### DISCUSSION TITLE
Understanding CPU speeds

### DISCUSSION THREAD

**Discussion by stefs12** [BlitterStudio/amiberry#913] [Q&A] [answered]
**What im trying to do:**

Im trying to understand how CPU speeds are working in amiberry, and how to tune it for my PI400.

Im building an Amiga System based on Workbench 3.2 and to meet most of the emulation needs of tasks I want to perform in a single config. In short i try to recreate an authentic Amiga experience more than have the most optimal config for every single game.

I am quite happy with the possibilities amiberry offers. I am trying though to understand a little more about CPU speeds. For now I see that the code (cfgfile.cpp) parses for different options. cpu_speed and finegrain_cpu_speed seem to be synonymous. cpu_frequency seems only to be relevant where cycle exact emulation is enabled (and even possible?).

The choice of 25Mhz in the UI leads to finegrain_cpu_speed=128, and 14Mhz to 1024 and fastest to the option cpu_speed=max. I dont fully understand the scale.

**What I try to figure out:** 

I like to find the optimal cpu speed for my system. I set the emulation to 68040 with FPU and "more compatible" in hopes to get the best software compatibility. I overclocked my PI400 to the standard 2.2 Ghz. I try to avoid JIT, because i noticed a lot of rather mixed results over the range of Amiga software I use. 

In that mode an emulation of 25 Mhz leads to a load of roughly 65 to 70 percent of a single core. Depending on the task at hand also less. In My observation the System could handle some more. But when i use the "Fastest" option the core load goes to 100%, which is expected of course, but seems to produce inconsistencies, because in many games or demos I now get a lot of sound stutter where I get perfectly fine results with the 25Mhz option.

Is there any safe middle ground between "25Mhz" and "Fastest"? If so, how do i approach it? Just set the finegrain_cpu_speed to 64 seems to have an adverse effect.

**Comment by stefs12** [comment, marked as answer]
Ok, I found out the error of my ways. Fastest CPU option and "cpu idle" is the way to fine tune this setting.

**Reply by midwan** [reply]
I'm happy you found a solution that works for you. The CPU speed setting is a bit of a mix, between what WinUAE does and what older versions of UAE4Arm did. The setting configures how many cycles will be given to the CPU, and the values you see there don't really mean much themselves, as they are only meant to be picked up and change the cycle shifting accordingly (so changing those values and putting something new there will not work as you think).

This is an area I would like to change and bring closer to WinUAE, but it requires a lot of changes in newcpu.cpp and potential new issues along the way, so I've left it for later. WinUAE no longer has the notion of "Mhz" for speed, but instead relies on either "Fastest possible" or "Approximate A500/A1200 or cycle exact", and from then on it has some sliders to adjust multipliers etc.

**Reply by stefs12** [reply]
Thanks for the insight. It's an interesting subject. Seems to be a lot of different strategies over the years.

**Comment by stefs12** [comment]
For me thats an answered question :) Seems im not able to mark it as answered?

**Reply by midwan** [reply]
That was because it was posted in General, not Q&A. I moved it and marked the answer. :)
